Initial support for DuckDuckGo search

losingkeys opened the pull request:
Warning: this breaks (at least) Serbian (I think?) translations, as I
don't know Serbian and it wasn't as easy as copying around the proper
nouns (like it was with French).

Also: fixed the Google Web Search URL. `http://www.google.com/#q=` wasn't working for me. Replaced with the one DuckDuckGo sends you to when you use `!g` (e.g. https://duckduckgo.com/?q=!g%20test).

Also: updated English translations to indicate we were doing a search,
not necessarily from the web (uses the Google Search app if possible,
and I intend to add support for the DDG app in the future).

Still todo (mostly copied from app/src/main/java/fr/neamar/kiss/result/DuckduckgoSearchResult.java):
- Use DDG app if available
- Programatically grab !bang commands and parse them for names (maybe
  just on startup or manually? Hardcode?
- Add option (in the settings underneath the option that enables DDG)
  that uses apps for some bangs if installed

losingkeys commented on app/src/main/res/values-sr/strings.xml:
```diff
@@ -32,8 +31,11 @@
 
-    <string name="search_desc">Тражи на вебу помоћу Гугла</string>
-    <string name="search_name">Гугл претрага</string>
+    <string name="google_search_desc">FIXME Enable Google Web Search</string>
```
These will need to be fixed

Neamar replied:
Serbian translation was provided by @pejakm on #111, maybe he'll be able to help on this one?

losingkeys commented:
Also the settings could probably be radio buttons instead of check boxes... Though if we started adding other search providers maybe that wouldn't make as much sense.

losingkeys commented:
Another question: should search results stay in the history? I think so

Neamar commented:
(regarding the last question: let's open an issue aside to discuss this)

Neamar commented:
(great work on the code, it's not in a pretty shape but it looks like you managed to work around!)

Regarding the PR, i'm wondering if it wouldn't be simpler to let the user picks its search provider in the option instead of adding both?

That way, you can enable search, and then pick your search provider?

losingkeys commented:
Using radio buttons? Or a pop-up dialog after enabling search? I'll try to
refactor to radio buttons.
On Jul 6, 2015 10:00, "Matthieu Bacconnier" <notifications@github.com>
wrote:

> (great work on the code, it's not in a pretty shape but it looks like you
> managed to work around!)
>
> Regarding the PR, i'm wondering if it wouldn't be simpler to let the user
> picks its search provider in the option instead of adding both?
>
> That way, you can enable search, and then pick your search provider?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Neamar/KISS/pull/124#issuecomment-118882052>.
>

losingkeys commented:
712f078d9c3939d72693e7d5e3b7e63d67e7b23b lets you choose a search engine from a list. Some notes:
- There are more untranslated strings because of this.
- It doesn't show which search engine you've chosen. I'm not sure if this is a huge deal, or if there's a better way to do it... but I'm okay with the solution if you are.

losingkeys commented:
I've rebased on Neamar/master. This is ready for review again :smile:. I'm okay with the fact that the settings don't show you which search engine you're using, if you are. Other than that, I've left some todos in [DuckduckgoSearchResult.java](https://github.com/losingkeys/KISS/blob/f01db60e53bb1b144a2968ddc6dff2037027812b/app/src/main/java/fr/neamar/kiss/result/DuckduckgoSearchResult.java#L38-L40), but I feel like those can be left for later (they're nice to have, not necessary for this to function.

The only thing left to do is code review (if you're okay with how it works) and translations. I've added the word 'FIXME' near untranslated strings, so they should be easy to find.

Neamar commented:
Hey! Great, thanks. I'll read this soon :-)

On Sat, Jul 25, 2015, 19:04 Josh Hoff <notifications@github.com> wrote:

> I've rebased on Neamar/master. This is ready for review again [image:
> :smile:]. I'm okay with the fact that the settings don't show you which
> search engine you're using, if you are. Other than that, I've left some
> todos in DuckduckgoSearchResult.java
> <https://github.com/losingkeys/KISS/blob/f01db60e53bb1b144a2968ddc6dff2037027812b/app/src/main/java/fr/neamar/kiss/result/DuckduckgoSearchResult.java#L38-L40>,
> but I feel like those can be left for later (they're nice to have, not
> necessary for this to function.
>
> The only thing left to do is code review (if you're okay with how it
> works) and translations. I've added the word 'FIXME' near untranslated
> strings, so they should be easy to find.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Neamar/KISS/pull/124#issuecomment-124862342>.
>

Neamar commented on app/src/main/java/fr/neamar/kiss/DataHandler.java:
```diff
@@ -57,9 +58,13 @@ public DataHandler(Context context) {
-
-        if (prefs.getBoolean("enable-search", true)) {
-            providers.add(new SearchProvider(context));
+        switch(prefs.getString("searchEngine", "none")) {
```
rather than none, i'd like to use "google".

Otherwise, when we enter a query without going into the settings, we get an empty screen :(

losingkeys replied:
Google is the default: https://github.com/losingkeys/KISS/blob/a55781007c234ee886fd2919ec94684453a03a7d/app/src/main/res/xml/preferences.xml#L49

Neamar commented on app/src/main/java/fr/neamar/kiss/result/DuckduckgoSearchResult.java:
```diff
@@ -0,0 +1,47 @@
+        return v;
+    }
+
+    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
```
is there anything requiring Honeycomb?

The app is targeted for Android < 3.0 too

losingkeys replied:
Not sure, I'll need to do some research and get back to you.

losingkeys replied:
Updated in 1955e86b1b5a37b17e6b2574585265971968897d

Neamar commented on app/src/main/res/values-fr/array.xml:
```diff
@@ -0,0 +1,13 @@
+        <item>none</item>
+    </string-array>
+    <string-array name="search_engine_name_list">
+        <item>FIXME DuckDuckGo</item>
```
DuckDuckGo

Neamar commented on app/src/main/res/values-fr/array.xml:
```diff
@@ -0,0 +1,13 @@
+    </string-array>
+    <string-array name="search_engine_name_list">
+        <item>FIXME DuckDuckGo</item>
+        <item>FIXME Google</item>
```
DuckDuckGo

Neamar commented on app/src/main/res/values-fr/array.xml:
```diff
@@ -0,0 +1,13 @@
+    <string-array name="search_engine_name_list">
+        <item>FIXME DuckDuckGo</item>
+        <item>FIXME Google</item>
+        <item>FIXME Disabled</item>
```
Aucun

Neamar commented on app/src/main/res/values-fr/strings.xml:
```diff
@@ -32,8 +31,8 @@
 
-    <string name="search_desc">Activer la recherche Google</string>
-    <string name="search_name">Recherche Google</string>
+    <string name="search_engines">FIXME Search Engines</string>
```
"Moteur de recherche"

Neamar commented on app/src/main/res/values-fr/strings.xml:
```diff
@@ -32,8 +31,8 @@
-    <string name="search_desc">Activer la recherche Google</string>
-    <string name="search_name">Recherche Google</string>
+    <string name="search_engines">FIXME Search Engines</string>
+    <string name="choose_search_engine">FIXME Choose a Search Engine</string>
```
"Choisir un moteur de recherche"

Neamar commented on app/src/main/res/xml/preferences.xml:
```diff
@@ -42,12 +42,13 @@
+        <ListPreference
+            android:title="@string/search_engines"
+            android:summary="@string/choose_search_engine"
+            android:key="searchEngine"
```
search-engine?

(i've used dash on other places, although i'm not sure of the Android convention on this one)

losingkeys replied:
android-studio was refusing to compile it with a dash. I'll try it again and get back to you.

Neamar commented:
Hey @losingkeys, just reviewed your PR with a few comments!

Looks good, however can we make "google" the default search engine (instead of "none")?

Final question, do you think we can merge as is, or do you want to implement your `//TODO` regarding DDG query parsing?

Neamar commented:
(also, sorry but due to #132 you'll need to pull master again :()

losingkeys commented:
I'll set the default to Google. I think we still need some translations (should we merge without them?). Also I'd like to show which search engine you've chosen on the settings page, but other than that I think the todo items can remain for future pull requests. I'll let you know when I'm ready for review again. Thanks for the code review.

Neamar commented:
Regarding the translation, I provided the french one, but no idea on the serbian... maybe we can leave them in plain english for now?

Ok for the settings, not sure if Android will allow this out of the bow -- you may need to customize the Preference on the activity `onCreate`?

Thanks for the submission :) i already have a few people willing to use your feature!

Nycroth commented:
:+1:

Neamar commented:
@losingkeys, i just spoke with @saveman71 regarding your PR and he asked a good question, is there any way we could use the standard Intent format to let the user pick an app?

You know, this kind of dialog:

![Android intent picker](http://i.stack.imgur.com/2agCh.jpg)

Currently, `SearchResult` is using:

```java
        Intent search = new Intent(Intent.ACTION_WEB_SEARCH);
        search.putExtra(SearchManager.QUERY, searchPojo.query)
```

Is there any other standard intent that would result in the "application picker" displaying both Google, DDG or *any other search engine the user has installed* ? This would make everything perfect :)

Rant on DDG app:
Actually, looking at the manifest, looks like DDG is using the `ASSIST` flag: http://stackoverflow.com/questions/14233330/replace-google-now-gesture, which may be a good starting point if we can send `ASSIST` a query -- not sure this can be done though

The interesting part is their manifest is sadly commented: https://github.com/duckduckgo/android/blob/fe0226fc7ad1940cc0e107cd9f6d33a88ab35031/AndroidManifest.xml#L40

Looks like @fgei disabled this a few month ago, not sure why. Maybe we could open an issue on their side?

losingkeys commented:
I've updated the PR to include the changes we discussed [earlier](https://github.com/Neamar/KISS/pull/124#issuecomment-127686839).

I like the idea of detecting installed search engine apps. I usually don't use a search engine app (I'd rather use a browser), so I'd prefer if this solution would support searching in a browser (e.g. "search using firefox").

Maybe it'd be even better if you could show all of them at once? For example, if you type in "some string that doesn't match any app names", then the KISS app would have these results:
- Amazon: "some string that doesn't match any app names"
- Google: "some string that doesn't match any app names"
- Firefox: "some string that doesn't match any app names"
- DuckDuckGo: "some string that doesn't match any app names"

These would depend on which apps you have installed. We could still detect the apps that support some type of search in the preferences, because you might want to disable one or more of them (e.g. if I installed an app for work or something I might not want to use it for searching).

Neamar commented:
Indeed. The question now is "how can we detect app with search functionality"?

I remember something about SearchProviders.

Seems the system already has some built-in tools for global search:

http://www.grokkingandroid.com/wordpress/wp-content/uploads/2012/07/search_settings_searchable_items.png

Maybe there's a way to list all applications with the Intent `android.app.SearchManager.INTENT_ACTION_GLOBAL_SEARCH` and use this lsit to create results? That would be pretty epic!

Reading the doc, there is http://developer.android.com/intl/zh-CN/reference/android/app/SearchManager.html#getSearchablesInGlobalSearch() which looks pretty useful.

losingkeys commented:
Closed in favor of #157.
